src/corelibs/U2View/src/ov_sequence/view_rendering/SequenceViewAnnotatedRenderer.cpp

```diff
@@ -188,16 +187,61 @@ void SequenceViewAnnotatedRenderer::drawAnnotation(QPainter& p, const QSize& can
                     const QString aText = prepareAnnotationText(aData, as);
                     drawBoundedText(p, annotationRect, aText);
                 }
-                if (simple &&
-                    annotationRect.width() > MIN_WIDTH_TO_DRAW_EXTRA_FEATURES &&
-                    displaySettings.displayCutSites) {
-                    drawCutSite(p, aData, r, annotationRect, as->color, canvasSize, visibleRange);
-                }
             }
             drawAnnotationConnections(p, a, as, displaySettings, canvasSize, visibleRange);
-        } else if (isRestrictionSite && simple && displaySettings.displayCutSites) {
-            p.setPen(borderPen);
-            drawCutSite(p, aData, r, QRect(), as->color, canvasSize, visibleRange);
+        }
+    }
+    // draw restriction sites
+    if (isRestrictionSite && displaySettings.displayCutSites) {
+        p.setPen(borderPen);
+        const bool isCircular = ctx->getSequenceObject()->isCircular();
+        auto seqLength = ctx->getSequenceLength();
+        if (isCircular && U1AnnotationUtils::isAnnotationContainsJunctionPoint(a, seqLength)) {
+            auto merged = U1AnnotationUtils::mergeAnnotatedRegionsAroundJunctionPoint(location, seqLength);
+            SAFE_POINT(merged.size() == 1, "Unexpected merged enzyme region", );
+
+            auto regionPair = merged.front();
+            // draw cuts separately,
+            // because they are located one on the 3' end of the direct sequence,
+            // and the other one on the 3' end of reverse-complementary sequence
+            const auto& qual =  aData->findFirstQualifierValue(GBFeatureUtils::QUALIFIER_CUT);
+            bool isDirectStrand = aData->getStrand().isDirect();
+            {
+                auto cutPosDirect = getCutPosition(regionPair.first, qual, isDirectStrand, true);
+                if (cutPosDirect != INCORRECT_CUT_POSITION) {
+                    auto annRect = getAnnotationRect(regionPair.first, canvasSize, visibleRange, selected, a, as);
+                    drawCutSite(p, aData, cutPosDirect, annRect, as->color, canvasSize, visibleRange, true);
+                }
+            }
+            {
+                auto cutPosComplement = getCutPosition(regionPair.second, qual, isDirectStrand, false);
+                if (cutPosComplement != INCORRECT_CUT_POSITION) {
+                    auto annRect = getAnnotationRect(regionPair.second, canvasSize, visibleRange, selected, a, as);
+                    drawCutSite(p, aData, cutPosComplement, annRect, as->color, canvasSize, visibleRange, false);
+                }
+            }
+        } else if (location.size() == 1) {
+            const auto& reg = location.first();
+            const auto& qual = aData->findFirstQualifierValue(GBFeatureUtils::QUALIFIER_CUT);
+            QRect annRect;
+            bool isDirectStrand = aData->getStrand().isDirect();\
+            {
```

Comment:
this nesting is not needed - it does not affect visibility scope of any variables or resolves any conflicts
If you think this is a more readable version - use comments to mark blocks instead

Reply:
OK